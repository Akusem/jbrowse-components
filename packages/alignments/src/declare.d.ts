declare module '@gmod/jbrowse-core/util/tracks'
declare module '@gmod/jbrowse-plugin-linear-genome-view'
declare module '@gmod/jbrowse-core/WiggleSNPAbstraction/util'
