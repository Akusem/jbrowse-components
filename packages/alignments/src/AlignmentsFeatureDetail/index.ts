import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import { ElementId } from '@gmod/jbrowse-core/util/types/mst'
import { types } from 'mobx-state-tree'

const configSchema = ConfigurationSchema('AlignmentsFeatureWidget', {})

const stateModel = types
  .model('AlignmentsFeatureWidget', {
    id: ElementId,
    type: types.literal('AlignmentsFeatureWidget'),
    featureData: types.frozen({}),
  })
  .actions(self => ({
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    setFeatureData(data: Record<string, any>) {
      self.featureData = data
    },
    clearFeatureData() {
      self.featureData = {}
    },
  }))

export { configSchema, stateModel }
export { default as ReactComponent } from './AlignmentsFeatureDetail'
