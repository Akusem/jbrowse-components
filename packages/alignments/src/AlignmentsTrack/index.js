export { default as ControlsReactComponent } from './components/TrackControls'
export { default as configSchemaFactory } from './configSchema'
export { default as modelFactory } from './model'
