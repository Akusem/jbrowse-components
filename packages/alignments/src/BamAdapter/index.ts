import PluginManager from '@gmod/jbrowse-core/PluginManager'
import configSchemaF from './configSchema'
import AdapterClass from './BamAdapter'

export default (pluginManager: PluginManager) => {
  return {
    configSchema: pluginManager.load(configSchemaF),
    AdapterClass,
  }
}
