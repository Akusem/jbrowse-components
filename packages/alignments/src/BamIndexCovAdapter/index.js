import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import { ObservableCreate } from '@gmod/jbrowse-core/util/rxjs'
import SimpleFeature from '@gmod/jbrowse-core/util/simpleFeature'
import {
  AdapterClass as BamAdapter,
  configSchema as bamAdapterConfigSchema,
} from '../BamAdapter'

export const configSchema = ConfigurationSchema(
  'BamIndexCovAdapter',
  {
    averageFeatureLength: {
      type: 'number',
      defaultValue: 150,
      description:
        'Index coverage produces number of features overlapping a region. If you want per-base feature depth, this can correspond to #features * #avg_len / 16kb',
    },
  },
  { baseConfiguration: bamAdapterConfigSchema, explicitlyTyped: true },
)

export class AdapterClass extends BamAdapter {
  static capabilities = ['getFeatures', 'getRefNames', 'getGlobalStats']

  private averageFeatureLength = 150

  constructor(opts) {
    super(opts)
    this.averageFeatureLength = opts.averageFeatureLength
  }

  async getGlobalStats() {
    return { scoreMax: 5000, scoreMin: 0 }
  }

  /**
   * Fetch features for a certain region. Use getFeaturesInRegion() if you also
   * want to verify that the store has features for the given reference sequence
   * before fetching.
   * @param {Region} param
   * @param {AbortSignal} [signal] optional signalling object for aborting the fetch
   * @returns {Observable[Feature]} Observable of Feature objects in the region
   */
  getFeatures({ refName, start, end }, { signal }) {
    return ObservableCreate(async observer => {
      await this.setup()
      const records = await this.bam.indexCov(refName, start, end, {
        signal,
      })
      console.log(this.averageFeatureLength)
      records.forEach(record => {
        record.score = record.score * this.averageFeatureLength / 16384
        observer.next(new SimpleFeature({ id: record.start + 1, data: record }))
      })
      observer.complete()
    })
  }
}
