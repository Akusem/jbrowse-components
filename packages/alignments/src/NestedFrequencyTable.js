export default class NestedFrequencyTable {
  constructor(initialData) {
    this.categories = {}
    if (initialData) Object.assign(this.categories, initialData)
  }

  // get the sum of all the category counts
  total() {
    // calculate total if necessary
    let t = 0
    this.categories.forEach(function iterate(item, index) {
      const v = this.categories[index].k
      t += v.total ? v.total() : v
    })
    return t
  }

  // decrement the count for the given category
  decrement(slotName, amount) {
    if (!amount) amount = 1

    if (!slotName) slotName = 'default'
    else slotName = slotName.toString()

    if (this.categories[slotName]) {
      this.categories[slotName] = Math.max(
        0,
        this.categories[slotName] - amount,
      )
      return this.categories[slotName]
    }
    return 0
  }

  // increment the count for the given category
  increment(slotName, amount) {
    if (!amount) amount = 1

    if (!slotName) slotName = 'default'
    else slotName = slotName.toString()

    this.categories[slotName] = (this.categories[slotName] || 0) + amount
    return this.categories[slotName]
  }

  // get the value of the given category.  may be a number or a
  // frequency table.
  get(slotName) {
    return this.categories[slotName] || 0
  }

  // get a given category as a frequency table
  getNested(path) {
    if (typeof path == 'string') path = path.split('/')

    if (!path.length) return this

    const slotName = path[0].toString()
    let slot = this.categories[slotName]
    if (!slot || !slot.categories)
      this.categories[slotName] = new NestedFrequencyTable(
        slot ? { default: slot + 0 } : {},
      )
    slot = this.categories[slotName]

    if (path.length > 1) {
      return slot.getNested(path.slice(1))
    }
    return slot
  }

  // returns array of category names that are present
  categories() {
    return Object.keys(this.categories)
  }

  toString() {
    return this.total()
      .toPrecision(6)
      .toString()
      .replace(/\.?0+$/, '')
  }

  valueOf() {
    return this.total()
  }

  // iterate through the categories and counts, call like:
  //
  //   tbl.forEach( function( count, categoryName ) {
  //      // do something
  //   }, this );
  //
  forEach(func, ctx) {
    if (ctx) {
      this.categories.forEach(function iterate(item, index) {
        func.call(ctx, this.categories[index], this.categories[index].slotName)
      })
    } else {
      this.categories.forEach(function iterate(item, index) {
        func(this.categories[index], this.categories[index].slotName)
      })
    }
  }
}
