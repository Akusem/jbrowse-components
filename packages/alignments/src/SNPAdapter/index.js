import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'

export { default as AdapterClass } from './SNPAdapter'

export const configSchema = ConfigurationSchema(
  'SNPAdapter',
  {
    subadapter: {
      type: 'frozen',
      defaultValue: {},
    },
  },
  { explicitlyTyped: true },
)

// export function configSchema(pluginManager) {
//   return ConfigurationSchema(
//     'SNPAdapter',
//     {
//       subadapter: pluginManager.pluggableConfigSchemaType('adapter'),
//     },
//     { explicitlyTyped: true },
//   )
// }
