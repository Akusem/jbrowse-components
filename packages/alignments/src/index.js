import AdapterType from '@gmod/jbrowse-core/pluggableElementTypes/AdapterType'
import DrawerWidgetType from '@gmod/jbrowse-core/pluggableElementTypes/DrawerWidgetType'
import TrackType from '@gmod/jbrowse-core/pluggableElementTypes/TrackType'
import Plugin from '@gmod/jbrowse-core/Plugin'
import { lazy } from 'react'
import {
  configSchema as alignmentsFeatureDetailConfigSchema,
  ReactComponent as AlignmentsFeatureDetailReactComponent,
  stateModel as alignmentsFeatureDetailStateModel,
} from './AlignmentsFeatureDetail'
import {
  configSchemaFactory as alignmentsTrackConfigSchemaFactory,
  modelFactory as alignmentsTrackModelFactory,
} from './AlignmentsTrack'
import {
  configSchemaFactory as snpTrackConfigSchemaFactory,
  modelFactory as snpTrackModelFactory,
} from './SNPTrack'
import {
  AdapterClass as SnpAdapterClass,
  configSchema as snpAdapterConfigSchema,
} from './SNPAdapter'
import {
  AdapterClass as BamAdapterClass,
  configSchema as bamAdapterConfigSchema,
} from './BamAdapter'
import {
  AdapterClass as CramAdapterClass,
  configSchema as cramAdapterConfigSchema,
} from './CramAdapter'
import PileupRenderer, {
  configSchema as pileupRendererConfigSchema,
  ReactComponent as PileupRendererReactComponent,
} from './PileupRenderer'
import SNPXYRenderer, {
  configSchema as SNPRendererConfigSchema,
  ReactComponent as SNPRendererReactComponent,
} from './SNPXYRenderer'

export default class extends Plugin {
  install(pluginManager) {
    pluginManager.addTrackType(() => {
      const configSchema = alignmentsTrackConfigSchemaFactory(pluginManager)
      return new TrackType({
        name: 'AlignmentsTrack',
        configSchema,
        stateModel: alignmentsTrackModelFactory(pluginManager, configSchema),
      })
    })

    pluginManager.addDrawerWidgetType(
      () =>
        new DrawerWidgetType({
          name: 'AlignmentsFeatureDrawerWidget',
          heading: 'Feature Details',
          configSchema: alignmentsFeatureDetailConfigSchema,
          stateModel: alignmentsFeatureDetailStateModel,
          LazyReactComponent: lazy(() => AlignmentsFeatureDetailReactComponent),
        }),
    )
    pluginManager.addTrackType(() => {
      const configSchema = snpTrackConfigSchemaFactory(pluginManager)
      return new TrackType({
        name: 'SNPTrack',
        configSchema,
        stateModel: snpTrackModelFactory(configSchema),
      })
    })
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'BamAdapter',
          configSchema: bamAdapterConfigSchema,
          AdapterClass: BamAdapterClass,
        }),
    )
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'SNPAdapter',
          configSchema: snpAdapterConfigSchema,
          AdapterClass: SnpAdapterClass,
        }),
    )
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'CramAdapter',
          requiresSequenceAdapter: true,
          configSchema: cramAdapterConfigSchema,
          AdapterClass: CramAdapterClass,
        }),
    )
    pluginManager.addRendererType(
      () =>
        new PileupRenderer({
          name: 'PileupRenderer',
          ReactComponent: PileupRendererReactComponent,
          configSchema: pileupRendererConfigSchema,
        }),
    )
    pluginManager.addRendererType(
      () =>
        new SNPXYRenderer({
          name: 'SNPXYRenderer',
          ReactComponent: SNPRendererReactComponent,
          configSchema: SNPRendererConfigSchema,
        }),
    )
  }
}
