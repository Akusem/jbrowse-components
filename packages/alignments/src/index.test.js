import { createTestSession } from '@gmod/jbrowse-web/src/jbrowseModel'
import { getSnapshot, types } from 'mobx-state-tree'

const createMockTrackStateModel = track =>
  types
    .model({
      name: 'testSession',
      selected: types.frozen(),
      bpPerPx: 0.05,
      staticBlocks: types.frozen(),
      track: track.stateModel,
      pluginManager: 'mockPluginManager',
    })
    .actions(self => {
      return {
        select(thing) {
          self.selection = thing
        },
        display() {},
      }
    })

const createMockTrack = track =>
  createMockTrackStateModel(track).create({
    staticBlocks: [],
    track: {
      configuration: track.configSchema.create({
        type: 'AlignmentsTrack',
      }),
      type: 'AlignmentsTrack',
    },
  })

test('create bam adapter config', () => {
  const { pluginManager } = createTestSession()

  const BamAdapter = pluginManager.getAdapterType('BamAdapter')
  const config = BamAdapter.configSchema.create({ type: 'BamAdapter' })
  expect(getSnapshot(config)).toMatchSnapshot({
    configId: expect.any(String),
    index: { configId: expect.any(String) },
  })
})
test('create track config', async () => {
  const { pluginManager } = createTestSession()

  const AlignmentsTrack = pluginManager.getTrackType('AlignmentsTrack')
  const config2 = AlignmentsTrack.configSchema.create({
    type: 'AlignmentsTrack',
  })
  expect(getSnapshot(config2)).toMatchSnapshot({
    adapter: {
      configId: expect.any(String),
      index: {
        configId: expect.any(String),
      },
    },
    configId: expect.any(String),
    renderers: {
      PileupRenderer: {
        configId: expect.any(String),
      },
      SvgFeatureRenderer: {
        configId: expect.any(String),
        labels: {
          configId: expect.any(String),
        },
      },
      configId: expect.any(String),
    },
  })
})

test('test selection in alignments track model with mock session', async () => {
  const { pluginManager } = createTestSession()

  const sessionModel = createMockTrack(
    pluginManager.getTrackType('AlignmentsTrack'),
  )

  sessionModel.track.selectFeature({
    id() {
      return 1234
    },
  })
  expect(sessionModel.selection.id()).toBe(1234)

  sessionModel.track.clearFeatureSelection()
  expect(sessionModel.selection).not.toBeTruthy()
})
