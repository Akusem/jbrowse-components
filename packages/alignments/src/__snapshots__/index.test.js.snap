// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create bam adapter config 1`] = `
Object {
  "bamLocation": Object {
    "base": "http://localhost/path/to/my.bam",
    "uri": "/path/to/my.bam",
  },
  "configId": Any<String>,
  "index": Object {
    "configId": Any<String>,
    "location": Object {
      "base": "http://localhost/path/to/my.bam.bai",
      "uri": "/path/to/my.bam.bai",
    },
  },
  "type": "BamAdapter",
}
`;

exports[`create track config 1`] = `
Object {
  "adapter": Object {
    "bamLocation": Object {
      "base": "http://localhost/path/to/my.bam",
      "uri": "/path/to/my.bam",
    },
    "configId": Any<String>,
    "index": Object {
      "configId": Any<String>,
      "location": Object {
        "base": "http://localhost/path/to/my.bam.bai",
        "uri": "/path/to/my.bam.bai",
      },
    },
    "type": "BamAdapter",
  },
  "configId": Any<String>,
  "renderers": Object {
    "PileupRenderer": Object {
      "configId": Any<String>,
      "type": "PileupRenderer",
    },
    "SvgFeatureRenderer": Object {
      "configId": Any<String>,
      "labels": Object {
        "configId": Any<String>,
      },
      "type": "SvgFeatureRenderer",
    },
    "configId": Any<String>,
  },
  "type": "AlignmentsTrack",
}
`;
